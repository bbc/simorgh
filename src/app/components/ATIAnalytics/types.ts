import { PageTypes, Platforms, Services } from '../../models/types/global';
import { RequestContextProps } from '../../contexts/RequestContext';
import { ServiceConfig } from '../../models/types/serviceConfig';

export interface AMPAnalyticsData {
  transport: {
    beacon: boolean;
    xhrpost: boolean;
    image: boolean;
  };
  requests: {
    base?: string;
    pageview: string;
  };
  triggers: { trackPageview: { on: string; request: string } };
}

export interface PageData {
  metadata?: {
    analyticsLabels?: {
      counterName?: string;
      contentId?: string;
      // eslint-disable-next-line camelcase
      nations_producer?: string;
      pageIdentifier?: string;
      pageTitle?: string;
      producer?: string;
    };
    atiAnalytics?: {
      producerId?: string;
      chapter?: string;
    };
    locators?: { optimoUrn?: string; curie?: string };
    passport?: {
      category?: { categoryId?: string; categoryName?: string };
      campaigns?: { campaignId?: string; campaignName?: string }[];
      language?: string;
    };
    tags?: {
      about?: {
        thingId?: string;
        thingLabel?: string;
        thingEnglishLabel?: string;
      }[];
    };
    id?: string;
    title?: string;
    language?: string;
    firstPublished?: number;
    lastPublished?: number;
    typ?: PageTypes;
  };
  promo?:
    | { headlines?: { seoHeadline?: string } }
    | { headlines?: { headline?: string } }
    | { subType: string; name: string; uri: string; id: string; type: string };
  id?: string;
  language?: string;
  pageIdentifier?: string;
  pageTitle?: string;
  firstRecordTimeStamp?: string;
  lastRecordTimeStamp?: string;
  contentType?: string;
  title?: string;
}

export interface ATIData {
  campaigns?: { campaignId?: string; campaignName?: string }[] | null;
  categoryName?: string | null;
  contentId?: string | null;
  contentType?: string;
  language?: string | null;
  ldpThingIds?: string | null;
  ldpThingLabels?: string | null;
  nationsProducer?: string | null;
  pageIdentifier?: string;
  pageTitle?: string | null;
  producerId?: string | null;
  producerName?: string | null;
  timePublished?: string | null;
  timeUpdated?: string | null;
}

export interface ATIDataWithContexts {
  requestContext: RequestContextProps;
  serviceContext: ServiceConfig;
  atiData: ATIData;
}

export interface ATIConfigurationDetailsProviders {
  requestContext: RequestContextProps;
  serviceContext: ServiceConfig;
  data?: PageData;
  atiData?: ATIData;
}

export interface ATIAnalyticsProps {
  baseUrl?: string;
  pageviewParams: string;
}

export interface ATIEventTrackingProps {
  campaignID?: string;
  componentName?: string;
  format?: string;
  pageIdentifier?: string;
  platform?: Platforms;
  producerId?: string;
  service?: Services;
  statsDestination?: string;
  type?: string;
  advertiserID?: string;
  url?: string;
}

export interface ATIPageTrackingProps {
  appName?: string;
  contentId?: string | null;
  contentType?: string | null;
  language?: string | null;
  ldpThingIds?: string | null;
  ldpThingLabels?: string | null;
  pageIdentifier?: string;
  pageTitle?: string | null;
  producerId?: string;
  libraryVersion?: string;
  platform?: Platforms;
  statsDestination?: string;
  timePublished?: string | null;
  timeUpdated?: string | null;
  origin?: string;
  previousPath?: string | null;
  categoryName?: string | null;
  campaigns?: { campaignId?: string; campaignName?: string }[] | null;
  nationsProducer?: string | null;
}

export interface ATIProps {
  data?: PageData;
  atiData?: ATIData;
}
