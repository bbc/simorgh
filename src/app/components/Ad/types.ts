import { PageTypes, Services, Direction } from '#app/models/types/global';

export type SlotType = 'leaderboard' | 'mpu';

export interface AdProps {
  slotType: SlotType;
  className?: string;
}

type SlotFunction = (slotType: string) => void;

interface Ads {
  registerSlot: SlotFunction;
  destroySlot: SlotFunction;
}

type PushFunction = (slotFunction: SlotFunction) => void;

interface Cmd {
  push: PushFunction;
}

export interface WindowWithDotCom {
  cmd: Cmd;
  ads: Ads;
}

export interface CanonicalAdBootstrapJsProps {
  adcampaign?: string | null;
}

export interface AdJsonDataProps {
  service: Services;
  slotType: SlotType;
  assetType: string;
}

export interface AmpAdSlotProps {
  service: Services;
  slotType: SlotType;
  pageType: PageTypes;
}

export interface AmpAdContentProps extends AmpAdSlotProps {
  dir: Direction;
  label: string;
}

export interface AmpAdProps extends AmpAdContentProps {
  ariaLabel: string;
}
