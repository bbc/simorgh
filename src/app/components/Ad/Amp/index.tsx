/** @jsxRuntime classic */
/** @jsx jsx */
import { jsx } from '@emotion/react';
import React, { useContext } from 'react';
import { Helmet } from 'react-helmet';
import { RequestContext } from '#contexts/RequestContext';
import {
  AMP_ACCESS_JS,
  AMP_ADS_JS,
} from '#psammead/psammead-assets/src/amp-boilerplate';
import { Services } from '#app/models/types/global';
import { ServiceContext } from '../../../contexts/ServiceContext';
import getAdsAriaLabel from '../utilities/getAdsAriaLabel';
import AdSlot from './AdSlot';
import styles from './index.styles';
import { AmpAdContentProps, AmpAdProps, AdProps } from '../types';

const AMP_ACCESS_DATA = (endpoint: string) => ({
  authorization: endpoint,
  noPingback: true,
  authorizationFallbackResponse: {
    error: true,
  },
});

const LABEL_LINK = 'https://www.bbc.com/usingthebbc/cookies/';

export const AMP_ACCESS_FETCH = (service: Services) => {
  const togglesEndpoint = `${process.env.SIMORGH_CONFIG_URL}?application=simorgh&service=${service}`;

  return (
    <script id="amp-access" type="application/json">
      {JSON.stringify(AMP_ACCESS_DATA(togglesEndpoint))}
    </script>
  );
};

const AdContent = ({
  service,
  dir,
  label,
  slotType,
  pageType,
}: AmpAdContentProps) => {
  const linkCss = [
    styles.link,
    dir === 'ltr' ? styles.ltrLink : styles.rtlLink,
  ];

  return (
    <React.Fragment>
      <a css={linkCss} href={LABEL_LINK} tabIndex={-1}>
        {label}
      </a>
      <AdSlot service={service} slotType={slotType} pageType={pageType} />
    </React.Fragment>
  );
};

const AdWithoutPlaceholder = ({
  ariaLabel,
  slotType,
  service,
  dir,
  label,
  pageType,
}: AmpAdProps) => {
  return (
    <div
      css={styles.ampGeo}
      amp-access="toggles.ads.enabled"
      amp-access-hide="true"
    >
      <section
        css={styles.section}
        aria-label={ariaLabel}
        role="region"
        data-e2e="advertisement"
        aria-hidden="true"
      >
        <div css={slotType === 'mpu' ? styles.mpu : styles.leaderboard}>
          <div css={styles.wrapper}>
            <AdContent
              service={service}
              dir={dir}
              label={label}
              pageType={pageType}
              slotType={slotType}
            />
          </div>
        </div>
      </section>
    </div>
  );
};

const AdWithPlaceholder = ({
  ariaLabel,
  slotType,
  service,
  dir,
  label,
  pageType,
}: AmpAdProps) => {
  return (
    <section
      css={styles.section}
      aria-label={ariaLabel}
      role="region"
      data-e2e="advertisement"
      aria-hidden="true"
    >
      <div css={slotType === 'mpu' ? styles.mpu : styles.leaderboard}>
        <div css={styles.wrapper}>
          <div
            css={styles.ampGeo}
            amp-access="toggles.ads.enabled"
            amp-access-hide="true"
          >
            <AdContent
              service={service}
              dir={dir}
              label={label}
              pageType={pageType}
              slotType={slotType}
            />
          </div>
        </div>
      </div>
    </section>
  );
};

const AmpAd = ({ slotType }: AdProps) => {
  const { translations, dir, service, showAdPlaceholder } =
    useContext(ServiceContext);
  const { pageType } = useContext(RequestContext);
  const label = translations?.ads?.advertisementLabel || 'Advertisement';

  const ariaLabel = getAdsAriaLabel(label, dir, slotType);

  const Advert = showAdPlaceholder ? AdWithPlaceholder : AdWithoutPlaceholder;

  return (
    <React.Fragment>
      <Helmet>
        {AMP_ADS_JS}
        {AMP_ACCESS_JS}
        {AMP_ACCESS_FETCH(service)}
      </Helmet>
      <Advert
        service={service}
        dir={dir}
        label={label}
        pageType={pageType}
        ariaLabel={ariaLabel}
        slotType={slotType}
      />
    </React.Fragment>
  );
};

export default AmpAd;
