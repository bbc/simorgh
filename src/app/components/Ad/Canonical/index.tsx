/** @jsx jsx */
/* @jsxFrag React.Fragment */
import { jsx } from '@emotion/react';
import React, { useEffect, useContext } from 'react';
import { Helmet } from 'react-helmet';
import { pathOr } from 'rambda';
import useLocation from '#hooks/useLocation';
import isLive from '../../../lib/utilities/isLive';
import useOperaMiniDetection from '../../../hooks/useOperaMiniDetection';
import { ServiceContext } from '../../../contexts/ServiceContext';
import getAdsAriaLabel from '../utilities/getAdsAriaLabel';
import adStyles from '../utilities/adSlot.styles';
import styles from './index.styles';
import { AdProps, SLOT_TYPES } from '../types';

const { LEADERBOARD } = SLOT_TYPES;

export const getBootstrapSrc = (queryString: string, useLegacy = false) => {
  const adsTestScript =
    'https://gn-web-assets.api.bbc.com/ngas/latest/test/dotcom-bootstrap.js';
  const adsLegacyTestScript =
    'https://gn-web-assets.api.bbc.com/ngas/latest/test/dotcom-bootstrap-legacy.js';
  const adsLiveScript =
    'https://gn-web-assets.api.bbc.com/ngas/latest/dotcom-bootstrap.js';
  const adsLegacyLiveScript =
    'https://gn-web-assets.api.bbc.com/ngas/latest/dotcom-bootstrap-legacy.js';

  const useLiveSrc =
    (isLive() && !queryString.includes('ads-test=true')) ||
    queryString.includes('ads-js-env=live');

  if (useLiveSrc) {
    return useLegacy ? adsLegacyLiveScript : adsLiveScript;
  }
  return useLegacy ? adsLegacyTestScript : adsTestScript;
};

const CanonicalAd = ({ slotType, className }: AdProps) => {
  const location = useLocation();
  const queryString = location.search;
  const { translations, dir } = useContext(ServiceContext);
  const label = pathOr(
    'Advertisement',
    ['ads', 'advertisementLabel'],
    translations,
  );
  const ariaLabel = getAdsAriaLabel({ label, dir, slotType });

  useEffect(() => {
    if (window.dotcom && location.href != null) {
      // @ts-expect-error  dotcom is added to the window object by BBC Ads script
      window.dotcom.cmd.push(() => {
        // @ts-expect-error  dotcom is added to the window object by BBC Ads script
        window.dotcom.ads.registerSlot(slotType);
      });
    }

    return () => {
      if (window.dotcom) {
        // @ts-expect-error  dotcom is added to the window object by BBC Ads script
        window.dotcom.cmd.push(() => {
          // @ts-expect-error  dotcom is added to the window object by BBC Ads script
          window.dotcom.ads.destroySlot(slotType);
        });
      }
    };
  }, [slotType, location]);

  const isOperaMini = useOperaMiniDetection();
  if (isOperaMini) {
    return null;
  }

  return (
    <>
      <Helmet>
        {/* Add Ad scripts to document head */}
        <script type="module" src={getBootstrapSrc(queryString)} async />
        <script noModule src={getBootstrapSrc(queryString, true)} async />
      </Helmet>

      <section
        css={[
          styles.section,
          slotType === LEADERBOARD ? adStyles.leaderboard : adStyles.mpu,
        ]}
        aria-label={ariaLabel}
        aria-hidden="true"
        role="region"
        data-e2e="advertisement"
        className={className}
      >
        <div id={`dotcom-${slotType}`} className="dotcom-ad" />
      </section>
    </>
  );
};

export default CanonicalAd;
