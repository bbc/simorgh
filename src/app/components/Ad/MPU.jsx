import React from 'react';
import styled from '@emotion/styled';
import { NEGATIVE_MARGIN } from '#lib/styles.const';
import { GEL_GROUP_1_SCREEN_WIDTH_MAX } from '#psammead/gel-foundations/src/breakpoints';
import BaseAdContainer from '.';

const AdContainer = styled(BaseAdContainer)`
  ${NEGATIVE_MARGIN}
  /* MPU gets misaligned with the original padding at smaller breakpoints */
  @media (max-width: ${GEL_GROUP_1_SCREEN_WIDTH_MAX}) {
    padding: 0;
  }
`;

export default props => {
  return <AdContainer slotType="mpu" {...props} />;
};
