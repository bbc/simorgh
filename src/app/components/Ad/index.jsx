import React, { useContext } from 'react';
import { oneOf, string } from 'prop-types';
import useToggle from '#hooks/useToggle';
import { RequestContext } from '#contexts/RequestContext';
import AmpAd from './Amp';
import CanonicalAd from './Canonical';

const AdContainer = ({ slotType, className }) => {
  const { isAmp } = useContext(RequestContext);
  const { enabled: adsEnabled } = useToggle('ads');

  if (!adsEnabled) {
    return null;
  }

  const Ad = isAmp ? AmpAd : CanonicalAd;
  return <Ad slotType={slotType} className={className} />;
};

AdContainer.propTypes = {
  slotType: oneOf(['leaderboard', 'mpu']).isRequired,
  className: string,
};

AdContainer.defaultProps = {
  className: null,
};

export default AdContainer;
