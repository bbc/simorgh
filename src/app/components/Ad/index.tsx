import React, { useContext } from 'react';
import useToggle from '#hooks/useToggle';
import { RequestContext } from '#contexts/RequestContext';
import AmpAd from './Amp';
import CanonicalAd from './Canonical';
import { AdProps } from './types';
import AdBootstrapJs from './Canonical/AdBootstrapJs';

const AdContainer = ({ slotType, className, adcampaign }: AdProps) => {
  const { isAmp, showAdsBasedOnLocation } = useContext(RequestContext);
  const { enabled: adsEnabled } = useToggle('ads');

  // Ads component will only be displayed if ads toggle is true and if showAdsBasedOnLocation is true
  if ([adsEnabled, showAdsBasedOnLocation].every(Boolean)) {
    const Ad = isAmp ? AmpAd : CanonicalAd;

    return (
      <>
        {/* dotcom and dotcomConfig need to be setup before the main dotcom javascript file is loaded */}
        {!isAmp && <AdBootstrapJs adcampaign={adcampaign} />}
        <Ad slotType={slotType} className={className} />
      </>
    );
  }
  return null;
};

export default AdContainer;
