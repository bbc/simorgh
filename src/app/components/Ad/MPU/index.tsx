/** @jsxRuntime classic */
/** @jsx jsx */
import { jsx } from '@emotion/react';
import Ad from '..';
import { AdProps } from '../types';
import styles from './index.styles';

export default ({ className }: AdProps) => {
  return <Ad css={styles.mpu} slotType="mpu" className={className} />;
};
