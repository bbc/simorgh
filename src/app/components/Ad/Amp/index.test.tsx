import { Services } from '#app/models/types/global';
import React from 'react';
import Ad, { AMP_ACCESS_FETCH } from '.';
import { render } from '../../react-testing-library-with-providers';
import { SlotType } from '../types';

const adJsonAttributes = (slotType: SlotType) => ({
  targeting: {
    slot: slotType,
    asset_type: 'index',
    channel: 'afrique',
  },
});

interface RenderParams {
  service: Services;
  showAdsBasedOnLocation: boolean;
}

const renderOptions: RenderParams = {
  service: 'afrique',
  showAdsBasedOnLocation: true,
};

describe('AMP Ads', () => {
  beforeAll(() => {
    process.env.SIMORGH_CONFIG_URL = 'https://mock-toggles-endpoint.bbc.co.uk';
  });

  afterAll(() => {
    delete process.env.SIMORGH_CONFIG_URL;
  });

  describe('Snapshots', () => {
    it('should correctly render an AMP leaderboard ad', () => {
      const { container } = render(<Ad slotType="leaderboard" />);
      expect(container).toMatchSnapshot();
    });

    it('should correctly render an AMP mpu ad', () => {
      const { container } = render(<Ad slotType="mpu" />, { ...renderOptions });
      expect(container).toMatchSnapshot();
    });
  });

  describe('Assertions', () => {
    it('should not render ad placeholder in UK when showAdPlaceholder in service config is false', () => {
      const { getByLabelText } = render(
        <div className="amp-geo-group-eea amp-geo-group-gbOrUnknown">
          <Ad slotType="leaderboard" />
        </div>,
        { ...renderOptions, service: 'news' },
      );

      expect(getByLabelText('Publicités')).not.toBeVisible();
    });

    it('should render ad placeholder in UK when showAdPlaceholder in service config is true', () => {
      const { getByLabelText } = render(
        <div className="amp-geo-group-eea amp-geo-group-gbOrUnknown">
          <Ad slotType="leaderboard" />
        </div>,
        { ...renderOptions },
      );

      expect(getByLabelText('Publicités')).toBeVisible();
    });

    it.each`
      location           | className
      ${'europe'}        | ${'amp-geo-group-eea'}
      ${'rest of world'} | ${''}
    `(
      'should render ad placeholder outside of the UK - $location',
      ({ className }) => {
        const { getByLabelText } = render(
          <div className={className}>
            <Ad slotType="leaderboard" />
          </div>,
          { ...renderOptions },
        );

        expect(getByLabelText('Publicités')).toBeVisible();
      },
    );

    it('should render two leaderboard ads', () => {
      const { container } = render(<Ad slotType="leaderboard" />, {
        ...renderOptions,
      });
      const ampAd = container.querySelectorAll('amp-ad');

      expect(ampAd.length).toBe(2);
    });

    it('should display leaderboard ad with values for all of the needed attributes', () => {
      const { container } = render(<Ad slotType="leaderboard" />, {
        ...renderOptions,
      });

      const ampAd = container.querySelectorAll('amp-ad');
      ampAd.forEach(ad => {
        expect(ad).toHaveAttribute('data-block-on-consent', 'default');
        expect(ad).toHaveAttribute('data-npa-on-unknown-consent', 'true');
        expect(ad).toHaveAttribute('media');
        expect(ad).toHaveAttribute('type');
        expect(ad).toHaveAttribute('width');
        expect(ad).toHaveAttribute('height');
        expect(ad).toHaveAttribute('data-multi-size');
        expect(ad).toHaveAttribute('data-slot');
        expect(ad).toHaveAttribute('data-amp-slot-index');
        expect(ad).toHaveAttribute('data-a4a-upgrade-type');
        expect(ad).toHaveAttribute(
          'json',
          JSON.stringify(adJsonAttributes('leaderboard')),
        );
      });
    });

    it('should render one mpu ad', () => {
      const { container } = render(<Ad slotType="mpu" />, {
        ...renderOptions,
      });
      const ampAd = container.querySelectorAll('amp-ad');

      expect(ampAd.length).toBe(1);
    });

    it('should display mpu ad with values for all of the needed attributes', () => {
      const { container } = render(<Ad slotType="mpu" />, {
        ...renderOptions,
      });

      const ampAd = container.querySelectorAll('amp-ad');
      ampAd.forEach(ad => {
        expect(ad).toHaveAttribute('data-block-on-consent', 'default');
        expect(ad).toHaveAttribute('data-npa-on-unknown-consent', 'true');
        expect(ad).toHaveAttribute('type');
        expect(ad).toHaveAttribute('width');
        expect(ad).toHaveAttribute('height');
        expect(ad).toHaveAttribute('data-multi-size');
        expect(ad).toHaveAttribute('data-slot');
        expect(ad).toHaveAttribute('data-amp-slot-index');
        expect(ad).toHaveAttribute('data-a4a-upgrade-type');
        expect(ad).toHaveAttribute(
          'json',
          JSON.stringify(adJsonAttributes('mpu')),
        );
      });
    });

    it('should render an `advertisement` label', () => {
      const { container } = render(<Ad slotType="leaderboard" />, {
        ...renderOptions,
      });
      const links = container.querySelectorAll('a');
      const advertisementLabel = links && links[0];
      expect(advertisementLabel.textContent).toEqual('Publicités');
      expect(advertisementLabel).toHaveAttribute('tabIndex', '-1');
    });
  });

  describe('AMP_ACCESS_FETCH', () => {
    it('should retrieve data from the correct endpoint', () => {
      const ampAccessFetch = jest.fn().mockImplementation(AMP_ACCESS_FETCH);
      const ampAccessData = ampAccessFetch('afrique');
      const expectedReturn =
        'https://mock-toggles-endpoint.bbc.co.uk?application=simorgh&service=afrique';

      expect(ampAccessFetch).toHaveReturned();
      expect(ampAccessFetch).toHaveBeenCalledWith('afrique');
      expect(ampAccessData.type).toEqual('script');
      expect(JSON.parse(ampAccessData.props.children).authorization).toEqual(
        expectedReturn,
      );
    });
  });
});
