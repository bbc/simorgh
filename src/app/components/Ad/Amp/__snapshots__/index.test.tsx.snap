// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AMP Ads Snapshots should correctly render an AMP leaderboard ad 1`] = `
.emotion-0 {
  background-color: #E6E8EA;
}

.emotion-1 {
  display: none;
  visibility: hidden;
}

@media (min-width: 20rem) {
  .emotion-1 {
    min-height: 5.1875rem;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-box-pack: center;
    -ms-flex-pack: center;
    -webkit-justify-content: center;
    justify-content: center;
    visibility: visible;
    padding: 1rem 0;
  }
}

@media (min-width: 63rem) {
  .emotion-1 {
    min-height: 18.6875rem;
    padding: 1.5rem 0;
  }
}

.emotion-2 {
  margin: 0 auto;
  text-align: center;
}

.amp-geo-pending .emotion-3,
.amp-geo-group-gbOrUnknown .emotion-3 {
  display: none;
  visibility: hidden;
}

.emotion-4 {
  font-size: 0.75rem;
  line-height: 1rem;
  font-family: Helmet,Freesans,Helvetica,Arial,sans-serif;
  font-style: normal;
  font-weight: 400;
  color: #5A5A5A;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  display: block;
  padding: 0.5rem 0;
  text-align: right;
}

@media (min-width: 20rem) and (max-width: 37.4375rem) {
  .emotion-4 {
    font-size: 0.75rem;
    line-height: 1rem;
  }
}

@media (min-width: 37.5rem) {
  .emotion-4 {
    font-size: 0.75rem;
    line-height: 1rem;
  }
}

.emotion-4:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div>
  <section
    aria-hidden="true"
    aria-label="Publicités"
    class="emotion-0"
    data-e2e="advertisement"
    role="region"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        <div
          amp-access="toggles.ads.enabled"
          amp-access-hide="true"
          class="emotion-3"
        >
          <a
            class="emotion-4"
            href="https://www.bbc.com/usingthebbc/cookies/"
            tabindex="-1"
          >
            Publicités
          </a>
          <amp-ad
            data-a4a-upgrade-type="amp-ad-network-doubleclick-impl"
            data-amp-slot-index="0"
            data-block-on-consent="default"
            data-multi-size="320x50,300x50"
            data-multi-size-validation="false"
            data-npa-on-unknown-consent="true"
            data-slot="/4817/bbccom.test.site.amp.news"
            height="50"
            json="{\\"targeting\\":{\\"slot\\":\\"leaderboard\\",\\"asset_type\\":\\"index\\",\\"channel\\":\\"afrique\\"}}"
            media="(max-width: 62.9375rem)"
            type="doubleclick"
            width="320"
          />
          <amp-ad
            data-a4a-upgrade-type="amp-ad-network-doubleclick-impl"
            data-amp-slot-index="0"
            data-block-on-consent="default"
            data-multi-size="970x250,728x90"
            data-multi-size-validation="false"
            data-npa-on-unknown-consent="true"
            data-slot="/4817/bbccom.test.site.amp.news"
            height="250"
            json="{\\"targeting\\":{\\"slot\\":\\"leaderboard\\",\\"asset_type\\":\\"index\\",\\"channel\\":\\"afrique\\"}}"
            media="(min-width: 63rem)"
            type="doubleclick"
            width="970"
          />
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`AMP Ads Snapshots should correctly render an AMP mpu ad 1`] = `
.emotion-0 {
  background-color: #E6E8EA;
}

.emotion-1 {
  display: none;
  visibility: hidden;
  padding: 0;
}

@media (min-width: 18.75rem) {
  .emotion-1 {
    min-height: 17.6875rem;
    padding: 1rem 0 0.5rem 0;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-box-pack: center;
    -ms-flex-pack: center;
    -webkit-justify-content: center;
    justify-content: center;
    visibility: visible;
  }
}

.emotion-2 {
  margin: 0 auto;
  text-align: center;
}

.amp-geo-pending .emotion-3,
.amp-geo-group-gbOrUnknown .emotion-3 {
  display: none;
  visibility: hidden;
}

.emotion-4 {
  font-size: 0.75rem;
  line-height: 1rem;
  font-family: Helmet,Freesans,Helvetica,Arial,sans-serif;
  font-style: normal;
  font-weight: 400;
  color: #5A5A5A;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  display: block;
  padding: 0.5rem 0;
  text-align: right;
}

@media (min-width: 20rem) and (max-width: 37.4375rem) {
  .emotion-4 {
    font-size: 0.75rem;
    line-height: 1rem;
  }
}

@media (min-width: 37.5rem) {
  .emotion-4 {
    font-size: 0.75rem;
    line-height: 1rem;
  }
}

.emotion-4:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div>
  <section
    aria-hidden="true"
    aria-label="Publicités 2"
    class="emotion-0"
    data-e2e="advertisement"
    role="region"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        <div
          amp-access="toggles.ads.enabled"
          amp-access-hide="true"
          class="emotion-3"
        >
          <a
            class="emotion-4"
            href="https://www.bbc.com/usingthebbc/cookies/"
            tabindex="-1"
          >
            Publicités
          </a>
          <amp-ad
            data-a4a-upgrade-type="amp-ad-network-doubleclick-impl"
            data-amp-slot-index="0"
            data-block-on-consent="default"
            data-multi-size="300x250"
            data-multi-size-validation="false"
            data-npa-on-unknown-consent="true"
            data-slot="/4817/bbccom.test.site.amp.news"
            height="250"
            json="{\\"targeting\\":{\\"slot\\":\\"mpu\\",\\"asset_type\\":\\"index\\",\\"channel\\":\\"afrique\\"}}"
            type="doubleclick"
            width="300"
          />
        </div>
      </div>
    </div>
  </section>
</div>
`;
